# Data Understanding & Cleaning - Olist Dataset

### 1) Import library

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import re

path = r"C:\Users\ADMIN\Downloads\Project\Brazillian E_Commerce\archive\\"
print(os.listdir(path))

['olist_customers_dataset.csv', 'olist_geolocation_dataset.csv', 'olist_orders_dataset.csv', 'olist_order_items_dataset.csv', 'olist_order_payments_dataset.csv', 'olist_order_reviews_dataset.csv', 'olist_products_dataset.csv', 'olist_sellers_dataset.csv', 'product_category_name_translation.csv']


### 2) Data Reading

In [2]:
df_customer = pd.read_csv(path + "olist_customers_dataset.csv")
df_geolocation = pd.read_csv(path + "olist_geolocation_dataset.csv")
df_orders = pd.read_csv(path + "olist_orders_dataset.csv")
df_order_items = pd.read_csv(path + "olist_order_items_dataset.csv")
df_order_payments = pd.read_csv(path + "olist_order_payments_dataset.csv")
df_order_reviews = pd.read_csv(path + "olist_order_reviews_dataset.csv")
df_products = pd.read_csv(path + "olist_products_dataset.csv")
df_sellers = pd.read_csv(path + "olist_sellers_dataset.csv")
df_cat = pd.read_csv(path + "product_category_name_translation.csv")

### 3) Data Understanding

#### 3.1) df_customer

In [3]:
print(df_customer.shape)
df_customer.head(2)

(99441, 5)


customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP

#### df_customer: Bảng này chứa thông tin về khách hàng.  Mỗi hàng trong bảng đại diện cho một khách hàng, với mỗi khách hàng được xác định bởi các thuộc tính hoặc tính năng được biểu thị bởi các cột trong bảng.
- customer_id : ID duy nhất cho mỗi giao dịch hoặc tương tác của khách hàng.
- customer_unique_id: ID duy nhất cho mỗi khách hàng.
- customer_zip_code_prefix: Mã bưu điện cho địa chỉ của khách hàng
- customer_city: Thành phó của khách hàng
- customer_state: Tiểu bang của khách hàng

#### 3.2) df_geolocation

In [4]:
print('Geolocation shape:', df_geolocation.shape)
df_geolocation.head()

Geolocation shape: (1000163, 5)


geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

#### Bảng này chứa dữ liệu xác định vị trí địa lý trong thế giới thực hoặc vị trí vật lý của các đối tượng khác nhau thông qua tọa độ GPS.  Mỗi hàng đại diện cho một vị trí địa lý duy nhất.
- geolocation_zip_code_prefix: Brazilian zip code (CEP) for a specific area
- geolocation_lat: Vĩ độ
- geolocation_lng: Kinh độ
- geolocation_city: Tên thành phố
- geolocation_state: Tên tiểu bang

#### 3.3) df_orders

In [5]:
print('Orders shape:', df_orders.shape)
df_orders['order_delivered_carrier_date'].unique()

Orders shape: (99441, 8)


array(['2017-10-04 19:55:00', '2018-07-26 14:31:00',
       '2018-08-08 13:50:00', ..., '2017-08-28 20:52:26',
       '2018-01-12 15:35:03', '2018-03-09 22:11:59'], dtype=object)

#### Bảng này chứa thông tin liên quan đến các đơn đặt hàng được đặt bởi khách hàng trên các nền tảng thương mại điện tử.  Mỗi hàng đại diện cho một đơn đặt hàng duy nhất từ  khách hàng.
- order_id: Định danh duy nhất cho mỗi Đơn đặt hàng
- customer_id: Định danh duy nhất cho khách hàng
- order_status: Trạng thái hiện tại của đơn hàng
- order_purchase_timestamp: Thời gian đặt hàng
- order_approved_at: Thời gian khi đơn hàng được duyệt
- order_delivered_carrier_date: Ngày đơn hàng được bàn giao cho đơn vị vận chuyển
- order_delivered_customer_date: Ngày giao hàng
- order_estimated_delivery_date: Ngày dự kiến giao hàng

#### 3.4) df_order_items

In [6]:
print('Order_items shape:', df_order_items.shape)
df_order_items.head()

Order_items shape: (112650, 7)


order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

#### Bảng chứa thông tin về các mục được đặt hàng trong hệ thống thương mại điện tử. Mỗi hàng đại diện cho một mục trong một đơn đặt hàng.
- order_id: Định danh duy nhất cho mỗi Đơn đặt hàng.
- order_item_id: Định danh từng mục trong 1 đơn đặt hàng
- product_id: Định danh duy nhất cho sản phẩm
- seller_id: Định danh duy nhất cho người bán hàng
- shipping_limit_date: Ngày hạn chót để sản phẩm được vận chuyển
- price: Giá bán của sản phẩm
- freight_value: Chi phí vận chuyển cho sản phẩm

#### 3.5) df_order_payments

In [7]:
print('Order_payments shape:', df_order_payments.shape)
df_order_payments.head()

Order_payments shape: (103886, 5)


order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

#### Bảng này chứa thông tin về các khoản thanh toán được thực hiện cho các đơn đặt hàng trong thương mại điện tử. Mỗi hàng đại diện cho một giao dịch thanh toán duy nhất, chi tiết cách khách hàng trả tiền cho đơn đặt hàng của họ.
- order_id: Định danh duy nhất cho mỗi Đơn đặt hàng.
- payment_sequential: Số thứ tự nếu có nhiều khoản thanh toán cho một đơn đặt hàng
- payment_type: Phương thức thanh toán (ví dụ: thẻ tín dụng, boleto, chứng từ)
- payment_installments: Số lần thanh toán với thẻ tín dụng
- payment_value: Tổng giá trị được trả cho đơn đặt hàng

#### 3.6) df_order_reviews

In [8]:
print('Order_reviews shape:', df_order_reviews.shape)
df_order_reviews.head()

Order_reviews shape: (99224, 7)


review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

#### Bảng này chứa thông tin về các đánh giá do khách hàng để lại cho các đơn đặt hàng mà họ đặt trên nền tảng thương mại điện tử. Mỗi hàng đại diện cho một phản hồi của khách hàng về đơn đặt hàng của họ.
- review_id: Định danh duy nhất cho mỗi đánh giá
- order_id: Định danh duy nhất cho mỗi Đơn đặt hàng.
- review_score: Điểm đánh giá được đưa ra bởi khách hàng, tỷ lệ từ 1 đến 5
- review_comment_title: Tiêu đề của nhận xét đánh giá, nếu có
- review_comment_message: Nhận xét chi tiết hoặc tin nhắn do khách hàng để lại như một phần của đánh giá
- review_creation_date: Ngày và giờ khi đánh giá được tạo
- review_answer_timestamp: Dấu thời gian khi đánh giá được người bán hoặc nền tảng phản hồi

#### 3.7) df_products

In [9]:
print('Products shape:', df_products.shape)
df_products.head()

Products shape: (32951, 9)


product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

#### Bảng chứa bộ sưu tập dữ liệu sản phẩm từ các nền tảng thương mại điện tử có chứa thông tin về các sản phẩm khác nhau như mô tả sản phẩm hoặc kích thước.
- product_id: Định danh duy nhất cho mỗi sản phẩm
- product_category_name: Danh mục mà sản phẩm thuộc về
- product_name_lenght: chiều dài của tên sản phẩm
- product_description_lenght: Chiều dài của mô tả sản phẩm
- product_photos_qty: Số lượng ảnh có sẵn cho sản phẩm
- product_weight_g: Trọng lượng của sản phẩm tính bằng gram
- product_length_cm: Chiều dài của sản phẩm tính bằng cm
- product_width_cm: Chiều rộng của sản phẩm tính bằng cm

#### 3.8) df_sellers

In [10]:
print('Sellers shape:', df_sellers.shape)
df_sellers.head()

Sellers shape: (3095, 4)


seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

#### Bảng này chứa thông tin về người bán đã đăng ký trên nền tảng thương mại điện tử. Mỗi hàng cung cấp chi tiết về một người bán duy nhất
- seller_id: Mã định danh duy nhất cho mỗi người bán
- seller_zip_code_prefix: Tiền tố mã zip cho địa chỉ của người bán
- seller_city: Thành phố của người bán
- seller_state: TIểu bang của người bán

#### 3.8) df_cat

In [11]:
print('Cat shape:', df_cat.shape)
df_cat.head()

Cat shape: (71, 2)


product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

#### Bảng chứa dữ liệu dịch tên loại sản phẩm, được sử dụng để dịch tên danh mục sản phẩm từ Bồ Đào Nha sang tiếng Anh
- Tên danh mục gốc bằng tiếng Bồ Đào Nha
- Tên danh mục được dịch bằng tiếng Anh

### 4) Data Cleaning

In [12]:
df_products.isnull().mean().values*100

array([0.        , 1.85123365, 1.85123365, 1.85123365, 1.85123365,
       0.00606962, 0.00606962, 0.00606962, 0.00606962])

In [13]:
def overview_data(data):
    # Create a DataFrame 'overview_data' with columns to capture various data characteristics
    overview_data = pd.DataFrame({
        'feature': data.columns.values,  # Column names
        'data_type': data.dtypes.values,  # Data types of columns
        'null_value(%)': data.isna().mean().values * 100,  # Percentage of missing values
        'neg_value(%)': [len(data[col][data[col] < 0]) / len(data) * 100 if col in data.select_dtypes(include=[np.number]).columns else 0 for col in data.columns],  # Percentage of negative values for numeric columns
        '0_value(%)': [len(data[col][data[col] == 0]) / len(data) * 100 if col in data.select_dtypes(include=[np.number]).columns else 0 for col in data.columns],  # Percentage of zero values for numeric columns
        'duplicate': data.duplicated().sum(),  # Number of duplicated rows
        'n_unique': data.nunique().values,  # Number of unique values for each column
        'sample_unique': [data[col].unique() for col in data.columns]  # Sample unique values for each column
    })
    
    # Round the values in the summary DataFrame to 3 decimal places
    return overview_data.round(3)

In [14]:
# Filter invalid city names based on pattern
def filtered_city(data, col):
    pattern = re.compile("[^a-z\sA-Z0-9-\'+]")
    filtered_df = data[data[col].str.contains(pattern)]
    return filtered_df

#### 4.1) df_geolocation

In [15]:
df_geolocation

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                               1037       -23.545621       -46.639292   
1                               1046       -23.546081       -46.644820   
2                               1046       -23.546129       -46.642951   
3                               1041       -23.544392       -46.639499   
4                               1035       -23.541578       -46.641607   
...                              ...              ...              ...   
1000158                        99950       -28.068639       -52.010705   
1000159                        99900       -27.877125       -52.224882   
1000160                        99950       -28.071855       -52.014716   
1000161                        99980       -28.388932       -51.846871   
1000162                        99950       -28.070104       -52.018658   

        geolocation_city geolocation_state  
0              sao paulo                SP  
1              sao paulo                SP  
2              sao paulo                SP  
3              sao paulo                SP  
4              sao paulo                SP  
...                  ...               ...  
1000158         tapejara                RS  
1000159   getulio vargas                RS  
1000160         tapejara                RS  
1000161  david canabarro                RS  
1000162         tapejara                RS  

[1000163 rows x 5 columns]

In [16]:
overview_data(df_geolocation)

feature data_type  null_value(%)  neg_value(%)  \
0  geolocation_zip_code_prefix     int64            0.0         0.000   
1              geolocation_lat   float64            0.0        99.866   
2              geolocation_lng   float64            0.0       100.000   
3             geolocation_city    object            0.0         0.000   
4            geolocation_state    object            0.0         0.000   

   0_value(%)  duplicate  n_unique  \
0         0.0     261831     19015   
1         0.0     261831    717360   
2         0.0     261831    717613   
3         0.0     261831      8011   
4         0.0     261831        27   

                                       sample_unique  
0  [1037, 1046, 1041, 1035, 1012, 1047, 1013, 102...  
1  [-23.54562128115268, -23.54608112703553, -23.5...  
2  [-46.63929204800168, -46.64482029837157, -46.6...  
3  [sao paulo, são paulo, sao bernardo do campo, ...  
4  [SP, RN, AC, RJ, ES, MG, BA, SE, PE, AL, PB, C...

- Bộ dữ liệu chứa hỗn hợp các loại dữ liệu. GeoLocation_zip_code_prefix thuộc loại int64, phù hợp với mã zip. GeoLocation_lat (vĩ độ) và GeoLocation_lng (kinh độ) là loại float64, phù hợp cho tọa độ địa lý. GeoLocation_City và GeoLocation_State thuộc loại chuỗi đại diện cho tên địa lý.
- Gần như toàn bộ giá trị âm 99% thuộc cột Long, 100% cột Lat chứa giá trị dương.
- Bộ dữ liệu có rất nhiều giá trị trùng lặp.

In [17]:
#Drop duplicate values
df_geolocation.drop_duplicates(inplace=True)
df_geolocation.duplicated().sum()

0

In [18]:
filtered_city(df_geolocation, 'geolocation_city')

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
5                               1012       -23.547762       -46.635361   
14                              1037       -23.545187       -46.637855   
17                              1024       -23.541390       -46.629899   
21                              1020       -23.552235       -46.628441   
22                              1011       -23.546690       -46.635447   
...                              ...              ...              ...   
1000094                        99940       -28.060955       -51.858637   
1000096                        99900       -27.884844       -52.230025   
1000098                        99900       -27.900022       -52.237668   
1000143                        99930       -27.913659       -52.248615   
1000147                        99900       -27.880973       -52.223206   

        geolocation_city geolocation_state  
5              são paulo                SP  
14             são paulo                SP  
17             são paulo                SP  
21             são paulo                SP  
22             são paulo                SP  
...                  ...               ...  
1000094           ibiaçá                RS  
1000096   getúlio vargas                RS  
1000098   getúlio vargas                RS  
1000143          estação                RS  
1000147   getúlio vargas                RS  

[63453 rows x 5 columns]

- Có 63453 dòng tại cột geolocation_city chứa ký tự đặc biệt

In [19]:
df_geolocation['geolocation_city'].value_counts()

geolocation_city
sao paulo               79927
rio de janeiro          35177
são paulo               19718
belo horizonte          19474
curitiba                11263
                        ...  
porto sao jose              1
riachão do bacamarte        1
riachao do bacamarte        1
algodão de jandaíra         1
luziapolis                  1
Name: count, Length: 8011, dtype: int64

In [20]:
# function to replace non-standard special characters with standard characters in a string
def replace_char(city_name):
    city_name = re.sub(r'[ãăâàáä]', 'a', city_name)
    city_name = re.sub(r'[íîì]', 'i', city_name)
    city_name = re.sub(r'[úûùü]', 'u', city_name)
    city_name = re.sub(r'[éêèë]', 'e', city_name)
    city_name = re.sub(r'[óõôòö]', 'o', city_name)
    city_name = re.sub(r'[ç]', 'c', city_name)
    return city_name

# Applying the function to clean 'geolocation_city' column
df_geolocation['geolocation_city'] = df_geolocation['geolocation_city'].apply(replace_char)

In [21]:
filtered_city(df_geolocation, 'geolocation_city')

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
106037                         4728       -23.639872       -46.713571   
301807                        13318       -23.293804       -45.959930   
302988                        13318       -23.256983       -47.055205   
308945                        13454       -22.734480       -47.380589   
377210                        17970       -21.269165       -51.668758   
421038                        21032       -22.857779       -43.252626   
491149                        25970       -22.392335       -42.989392   
502643                        27163       -22.467331       -43.829419   
506885                        27598       -22.411815       -44.553700   
522746                        28927       -22.622491       -42.004973   
524606                        28993       -22.897623       -42.474169   
525385                        28930       -22.969370       -42.029834   
530251                        29187       -20.041422       -40.187039   
600087                        35698       -19.952891       -44.705767   
622585                        36905       -20.192100       -42.125735   
681827                        42820       -12.623163       -38.088404   
695077                        45936       -18.006085       -39.869503   
695078                        45936       -18.012923       -39.853185   
698306                        47220        -9.517593       -43.014366   
726758                        57010        -9.674896       -35.756557   
823386                        78278       -15.316965       -58.006672   
851430                        81470       -25.571748       -49.334374   
896733                        87365       -24.271860       -53.069433   
896977                        87365       -24.281905       -53.074516   
897217                        87365       -24.277697       -53.074930   
900755                        88058       -27.450674       -48.401693   

                              geolocation_city geolocation_state  
106037                              sa£o paulo                SP  
301807                       jacare (cabreuva)                SP  
302988                       jacare (cabreuva)                SP  
308945                   santa barbara d`oeste                SP  
377210         sao joao do pau d%26apos%3balho                SP  
421038  rio de janeiro, rio de janeiro, brasil                RJ  
491149                            ´teresopolis                RJ  
502643    california da barra (barra do pirai)                RJ  
506885                       penedo (itatiaia)                RJ  
522746                     tamoios (cabo frio)                RJ  
524606           bacaxa (saquarema) - distrito                RJ  
525385                      ...arraial do cabo                RJ  
530251        praia grande (fundao) - distrito                ES  
600087                    antunes (igaratinga)                MG  
622585                      realeza (manhuacu)                MG  
681827       monte gordo (camacari) - distrito                BA  
695077                       itabatan (mucuri)                BA  
695078                       itabatan (mucuri)                BA  
698306  campo alegre de lourdes, bahia, brasil                BA  
726758                                 maceia³                AL  
823386                lambari d%26apos%3boeste                MT  
851430                                * cidade                PR  
896733                          4o. centenario                PR  
896977                           4º centenario                PR  
897217                           4º centenario                PR  
900755                    florian&oacute;polis                SC

In [22]:
df_geolocation[df_geolocation['geolocation_zip_code_prefix'] == 17970]

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
376805                        17970       -21.272372       -51.666023   
376827                        17970       -21.270330       -51.667482   
376905                        17970       -21.271680       -51.663581   
376937                        17970       -21.271582       -51.666313   
377022                        17970       -21.268891       -51.663565   
377073                        17970       -21.272542       -51.666665   
377082                        17970       -21.269781       -51.665765   
377092                        17970       -21.269704       -51.666216   
377138                        17970       -21.272594       -51.665497   
377186                        17970       -21.268528       -51.667841   
377210                        17970       -21.269165       -51.668758   
377306                        17970       -21.271869       -51.663261   
377324                        17970       -21.233559       -51.676964   
377331                        17970       -21.269781       -51.665765   
377496                        17970       -21.270829       -51.664587   

                       geolocation_city geolocation_state  
376805           sao joao do pau d'alho                SP  
376827           sao joao do pau d'alho                SP  
376905           sao joao do pau d'alho                SP  
376937           sao joao do pau d alho                SP  
377022           sao joao do pau d alho                SP  
377073           sao joao do pau d'alho                SP  
377082           sao joao do pau d'alho                SP  
377092            sao joao do pau dalho                SP  
377138           sao joao do pau d'alho                SP  
377186            sao joao do pau dalho                SP  
377210  sao joao do pau d%26apos%3balho                SP  
377306           sao joao do pau d alho                SP  
377324           sao joao do pau d'alho                SP  
377331           sao joao do pau d'alho                SP  
377496           sao joao do pau d'alho                SP

In [23]:
for i in df_geolocation['geolocation_zip_code_prefix'].unique():
    mode_city = df_geolocation[df_geolocation['geolocation_zip_code_prefix'] == i]['geolocation_city'].mode().values[0]
    df_geolocation.loc[df_geolocation['geolocation_zip_code_prefix'] == i, 'geolocation_city'] = mode_city

In [24]:
df_geolocation['geolocation_city'] = df_geolocation['geolocation_city'].str.title()
filtered_city(df_geolocation, 'geolocation_city')

Empty DataFrame
Columns: [geolocation_zip_code_prefix, geolocation_lat, geolocation_lng, geolocation_city, geolocation_state]
Index: []

- **df_geolocation has cleaned**

#### 4.2) df_customer

In [25]:
df_customer.head(5)

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [26]:
overview_data(df_customer)

feature data_type  null_value(%)  neg_value(%)  \
0               customer_id    object            0.0           0.0   
1        customer_unique_id    object            0.0           0.0   
2  customer_zip_code_prefix     int64            0.0           0.0   
3             customer_city    object            0.0           0.0   
4            customer_state    object            0.0           0.0   

   0_value(%)  duplicate  n_unique  \
0         0.0          0     99441   
1         0.0          0     96096   
2         0.0          0     14994   
3         0.0          0      4119   
4         0.0          0        27   

                                       sample_unique  
0  [06b8999e2fba1a1fbc88172c00ba8bc7, 18955e83d33...  
1  [861eff4711a542e4b93843c6dd7febb0, 290c77bc529...  
2  [14409, 9790, 1151, 8775, 13056, 89254, 4534, ...  
3  [franca, sao bernardo do campo, sao paulo, mog...  
4  [SP, SC, MG, PR, RJ, RS, PA, GO, ES, BA, MA, M...

- Có 2 kiểu dữ liệu trong bộ dữ liệu
    + customer_id, customer_unique_id, customer_city, customer_state: Kiểu string là phù hợp
    + customer_zip_code_prefix: kiểu int là phù hợp
    + Bộ dữ liệu không có giá trị âm hoặc bằng 0, không có giá trị trùng lặp, không giá trị null
    + Tuy nhiên có cột customer_city cần kiểm tra tính đồng nhất

In [27]:
filtered_city(df_customer, 'customer_city')

Empty DataFrame
Columns: [customer_id, customer_unique_id, customer_zip_code_prefix, customer_city, customer_state]
Index: []

In [28]:
# Đồng nhất cột customer_city với cột geolocation_city đã được làm sạch của df_geolocation.
df_customer = df_customer.merge(df_geolocation[['geolocation_city', 'geolocation_zip_code_prefix']],
                    left_on = df_customer['customer_zip_code_prefix'],
                    right_on = df_geolocation['geolocation_zip_code_prefix'],
                    how = 'left'
                 ).drop_duplicates()


In [29]:
df_customer['geolocation_city'].fillna(df_customer['customer_city'], inplace = True)
df_customer['geolocation_city'] = df_customer['geolocation_city'].str.title()
df_customer['customer_city'] = df_customer['geolocation_city']
df_customer.drop(columns=['geolocation_city', 'geolocation_zip_code_prefix'], inplace=True)

In [30]:
df_customer['customer_city'].value_counts()

customer_city
Sao Paulo               15540
Rio De Janeiro           6882
Belo Horizonte           2773
Brasilia                 2135
Curitiba                 1521
                        ...  
Central                     1
Nova Independencia          1
Cha De Alegria              1
Sao Joao Do Itaperiu        1
Eugenio De Castro           1
Name: count, Length: 4089, dtype: int64

#### 4.3) df_orders

In [31]:
overview_data(df_orders)

feature data_type  null_value(%)  neg_value(%)  \
0                       order_id    object          0.000             0   
1                    customer_id    object          0.000             0   
2                   order_status    object          0.000             0   
3       order_purchase_timestamp    object          0.000             0   
4              order_approved_at    object          0.161             0   
5   order_delivered_carrier_date    object          1.793             0   
6  order_delivered_customer_date    object          2.982             0   
7  order_estimated_delivery_date    object          0.000             0   

   0_value(%)  duplicate  n_unique  \
0           0          0     99441   
1           0          0     99441   
2           0          0         8   
3           0          0     98875   
4           0          0     90733   
5           0          0     81018   
6           0          0     95664   
7           0          0       459   

                                       sample_unique  
0  [e481f51cbdc54678b7cc49136f2d6af7, 53cdb2fc8bc...  
1  [9ef432eb6251297304e76186b10a928d, b0830fb4747...  
2  [delivered, invoiced, shipped, processing, una...  
3  [2017-10-02 10:56:33, 2018-07-24 20:41:37, 201...  
4  [2017-10-02 11:07:15, 2018-07-26 03:24:27, 201...  
5  [2017-10-04 19:55:00, 2018-07-26 14:31:00, 201...  
6  [2017-10-10 21:25:13, 2018-08-07 15:27:45, 201...  
7  [2017-10-18 00:00:00, 2018-08-13 00:00:00, 201...

- Các cột trong bộ dữ liệu hiện tại có kiểu dữ liệu chuỗi tuy nhiên có 5 cột: order_purchase_timestamp, order_approved_at, order_delivered_carrier_date, order_delivered_customer_date, order_estimated_delivery_date phải là kiểu thời gian.
- Có < 3% giá trị null tại 3 cột order_approved_at, order_delivered_carrier_date, order_delivered_customer_date có thể do đơn hàng mới được lên đơn hoặc chưa giao hàng..
- Không có giá trị âm, =0, duplicate
- Cần lọc dữ liệu theo logic thời gian, ví dụ ngày **order_approved_at** bắt buộc phải lớn hơn hoặc bằng so với ngày **order_purchase_timestamp**

In [32]:
col = ['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']
for c in col:
    df_orders[c] = pd.to_datetime(df_orders[c])

In [33]:
df_orders.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [34]:
print(f"Before: {df_orders.shape}")
df_orders = df_orders[(df_orders['order_approved_at'].isnull()) | ((df_orders['order_approved_at'].notnull()) & (df_orders['order_approved_at'] >= df_orders['order_purchase_timestamp']))]
df_orders = df_orders[(df_orders['order_delivered_carrier_date'].isnull()) | ((df_orders['order_delivered_carrier_date'].notnull()) & (df_orders['order_delivered_carrier_date'] >= df_orders['order_approved_at']))]
df_orders = df_orders[(df_orders['order_approved_at'].isnull()) | (df_orders['order_delivered_carrier_date'].isnull()) | (df_orders['order_delivered_customer_date'].isnull()) | ((df_orders['order_delivered_customer_date'].notnull()) & (df_orders['order_delivered_customer_date'] >= df_orders['order_delivered_carrier_date']))]
df_orders = df_orders[(df_orders['order_approved_at'].isnull()) | (df_orders['order_delivered_carrier_date'].isnull()) | (df_orders['order_delivered_customer_date'].isnull()) | (df_orders['order_estimated_delivery_date'].isnull()) | ((df_orders['order_estimated_delivery_date'].notnull()) & (df_orders['order_estimated_delivery_date'] >= df_orders['order_delivered_customer_date']))]
print(f"After: {df_orders.shape}")

Before: (99441, 8)
After: (90252, 8)


In [35]:
overview_data(df_orders)

feature       data_type  null_value(%)  neg_value(%)  \
0                       order_id          object          0.000             0   
1                    customer_id          object          0.000             0   
2                   order_status          object          0.000             0   
3       order_purchase_timestamp  datetime64[ns]          0.000             0   
4              order_approved_at  datetime64[ns]          0.162             0   
5   order_delivered_carrier_date  datetime64[ns]          1.976             0   
6  order_delivered_customer_date  datetime64[ns]          3.275             0   
7  order_estimated_delivery_date  datetime64[ns]          0.000             0   

   0_value(%)  duplicate  n_unique  \
0           0          0     90252   
1           0          0     90252   
2           0          0         8   
3           0          0     89789   
4           0          0     83262   
5           0          0     73952   
6           0          0     86599   
7           0          0       458   

                                       sample_unique  
0  [e481f51cbdc54678b7cc49136f2d6af7, 53cdb2fc8bc...  
1  [9ef432eb6251297304e76186b10a928d, b0830fb4747...  
2  [delivered, invoiced, shipped, processing, una...  
3  [2017-10-02 10:56:33, 2018-07-24 20:41:37, 201...  
4  [2017-10-02 11:07:15, 2018-07-26 03:24:27, 201...  
5  [2017-10-04 19:55:00, 2018-07-26 14:31:00, 201...  
6  [2017-10-10 21:25:13, 2018-08-07 15:27:45, 201...  
7  [2017-10-18 00:00:00, 2018-08-13 00:00:00, 201...

In [36]:
df_orders['order_status'] = df_orders['order_status'].str.title()
df_orders['order_status'].value_counts()

order_status
Delivered      87299
Shipped         1098
Canceled         624
Unavailable      609
Invoiced         314
Processing       301
Created            5
Approved           2
Name: count, dtype: int64

#### 4.4) df_order_items

In [37]:
overview_data(df_order_items)

feature data_type  null_value(%)  neg_value(%)  0_value(%)  \
0             order_id    object            0.0           0.0        0.00   
1        order_item_id     int64            0.0           0.0        0.00   
2           product_id    object            0.0           0.0        0.00   
3            seller_id    object            0.0           0.0        0.00   
4  shipping_limit_date    object            0.0           0.0        0.00   
5                price   float64            0.0           0.0        0.00   
6        freight_value   float64            0.0           0.0        0.34   

   duplicate  n_unique                                      sample_unique  
0          0     98666  [00010242fe8c5a6d1ba2dd792cb16214, 00018f77f2f...  
1          0        21  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...  
2          0     32951  [4244733e06e7ecb4970a6e2683c13e61, e5f2d52b802...  
3          0      3095  [48436dade18ac8b2bce089ec2a041202, dd7ddc04e1b...  
4          0     93318  [2017-09-19 09:45:35, 2017-05-03 11:05:13, 201...  
5          0      5968  [58.9, 239.9, 199.0, 12.99, 199.9, 21.9, 19.9,...  
6          0      6999  [13.29, 19.93, 17.87, 12.79, 18.14, 12.69, 11....

In [38]:
df_order_items.head(3)

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   

   shipping_limit_date  price  freight_value  
0  2017-09-19 09:45:35   58.9          13.29  
1  2017-05-03 11:05:13  239.9          19.93  
2  2018-01-18 14:48:30  199.0          17.87

- Bộ dữ liệu bao gồm các kiểu dữ liệu như order_id, product_id, seller_id kiểu Object là hợp lý, cột order_item_id: kiểu int là hợp lý, cột price, freight_value kiểu float là hợp lý.
- Cột shipping_limit_date hiện tại kiểu object là không hợp lý.
- Không có giá trị null, giá trị âm, dupplicate
- Có 0.34% giá trị bằng 0 tại cột freight_value có thể đây là những đơn hàng free ship

In [39]:
df_order_items['shipping_limit_date'] = df_order_items['shipping_limit_date'].astype('datetime64[ns]')
df_order_items.dtypes

order_id                       object
order_item_id                   int64
product_id                     object
seller_id                      object
shipping_limit_date    datetime64[ns]
price                         float64
freight_value                 float64
dtype: object

In [40]:
overview_data(df_order_payments)

feature data_type  null_value(%)  neg_value(%)  0_value(%)  \
0              order_id    object            0.0           0.0       0.000   
1    payment_sequential     int64            0.0           0.0       0.000   
2          payment_type    object            0.0           0.0       0.000   
3  payment_installments     int64            0.0           0.0       0.002   
4         payment_value   float64            0.0           0.0       0.009   

   duplicate  n_unique                                      sample_unique  
0          0     99440  [b81ef226f3fe1789b1e8b2acac839d17, a9810da8291...  
1          0        29  [1, 2, 4, 5, 3, 8, 6, 7, 10, 11, 17, 19, 27, 1...  
2          0         5  [credit_card, boleto, voucher, debit_card, not...  
3          0        24  [8, 1, 2, 3, 6, 5, 4, 10, 7, 12, 9, 13, 15, 24...  
4          0     29077  [99.33, 24.39, 65.71, 107.78, 128.45, 96.12, 8...

In [41]:
df_order_payments.head(3)

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71

- Kiểu dữ liệu của các cột là hợp lý
- Bộ dữ liệu không có giá trị null, duplicate, giá trị âm
- Có 1 số giá trị =0 tại 2 cột payment_installments và payment_value cần phải xem xét.

In [42]:
df_order_payments[df_order_payments['payment_installments']==0]

order_id  payment_sequential payment_type  \
46982  744bade1fcf9ff3f31d860ace076d422                   2  credit_card   
79014  1a57108394169c0b47d8f876acc9ba2d                   2  credit_card   

       payment_installments  payment_value  
46982                     0          58.69  
79014                     0         129.94

In [43]:
df_order_payments = df_order_payments[df_order_payments['payment_installments']!=0]

In [44]:
df_order_payments[df_order_payments['payment_value']==0]

order_id  payment_sequential payment_type  \
19922   8bcbe01d44d147f901cd3192671144db                   4      voucher   
36822   fa65dad1b0e818e3ccc5cb0e39231352                  14      voucher   
43744   6ccb433e00daae1283ccc956189c82ae                   4      voucher   
51280   4637ca194b6387e2d538dc89b124b0ee                   1  not_defined   
57411   00b1cb0320190ca0daa2c88b35206009                   1  not_defined   
62674   45ed6e85398a87c253db47c2d9f48216                   3      voucher   
77885   fa65dad1b0e818e3ccc5cb0e39231352                  13      voucher   
94427   c8c528189310eaa44a745b8d9d26908b                   1  not_defined   
100766  b23878b3e8eb4d25a158f57d96331b18                   4      voucher   

        payment_installments  payment_value  
19922                      1            0.0  
36822                      1            0.0  
43744                      1            0.0  
51280                      1            0.0  
57411                      1            0.0  
62674                      1            0.0  
77885                      1            0.0  
94427                      1            0.0  
100766                     1            0.0

In [45]:
notdefined_0 = df_order_payments[(df_order_payments['payment_value']==0) & (df_order_payments['payment_type']=='not_defined')]
df_order_payments.drop(notdefined_0.index, inplace=True)
df_order_payments[df_order_payments['payment_value']==0]

order_id  payment_sequential payment_type  \
19922   8bcbe01d44d147f901cd3192671144db                   4      voucher   
36822   fa65dad1b0e818e3ccc5cb0e39231352                  14      voucher   
43744   6ccb433e00daae1283ccc956189c82ae                   4      voucher   
62674   45ed6e85398a87c253db47c2d9f48216                   3      voucher   
77885   fa65dad1b0e818e3ccc5cb0e39231352                  13      voucher   
100766  b23878b3e8eb4d25a158f57d96331b18                   4      voucher   

        payment_installments  payment_value  
19922                      1            0.0  
36822                      1            0.0  
43744                      1            0.0  
62674                      1            0.0  
77885                      1            0.0  
100766                     1            0.0

In [46]:
overview_data(df_order_reviews)

feature data_type  null_value(%)  neg_value(%)  0_value(%)  \
0                review_id    object          0.000           0.0         0.0   
1                 order_id    object          0.000           0.0         0.0   
2             review_score     int64          0.000           0.0         0.0   
3     review_comment_title    object         88.342           0.0         0.0   
4   review_comment_message    object         58.703           0.0         0.0   
5     review_creation_date    object          0.000           0.0         0.0   
6  review_answer_timestamp    object          0.000           0.0         0.0   

   duplicate  n_unique                                      sample_unique  
0          0     98410  [7bc2406110b926393aa56f80a40eba40, 80e641a11e5...  
1          0     98673  [73fc7af87114b39712e6da79b0a377eb, a548910a1c6...  
2          0         5                                    [4, 5, 1, 3, 2]  
3          0      4527  [nan, recomendo, Super recomendo, Não chegou m...  
4          0     36159  [nan, Recebi bem antes do prazo estipulado., P...  
5          0       636  [2018-01-18 00:00:00, 2018-03-10 00:00:00, 201...  
6          0     98248  [2018-01-18 21:46:59, 2018-03-11 03:05:13, 201...

In [47]:
df_order_reviews.head(3)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   

   review_score review_comment_title review_comment_message  \
0             4                  NaN                    NaN   
1             5                  NaN                    NaN   
2             5                  NaN                    NaN   

  review_creation_date review_answer_timestamp  
0  2018-01-18 00:00:00     2018-01-18 21:46:59  
1  2018-03-10 00:00:00     2018-03-11 03:05:13  
2  2018-02-17 00:00:00     2018-02-18 14:36:24

- Kiểu dữ liệu của các cột review_id, order_id, review_comment_title, review_comment_message là object là phù hợp, review_score kiểu int là phù hợp
- review_creation_date, review_answer_timestamp là kiểu object ko phù hợp.
- Có nhiều giá trị null tại 2 cột review_comment_title, review_comment_message cũng có thể có nhiều khách hàng không để lại tiêu đề hoặc tin nhắn.
- Không có giá trị dup, giá trị số = 0 hoặc < 0.

In [48]:
#df_order_reviews.drop(columns=['review_comment_title', 'review_comment_message'], inplace = True )

In [49]:
cols = ['review_creation_date', 'review_answer_timestamp']
for c in cols:
    df_order_reviews[c] = pd.to_datetime(df_order_reviews[c])

In [50]:
df_order_reviews.dtypes

review_id                          object
order_id                           object
review_score                        int64
review_comment_title               object
review_comment_message             object
review_creation_date       datetime64[ns]
review_answer_timestamp    datetime64[ns]
dtype: object

In [51]:
overview_data(df_products)

feature data_type  null_value(%)  neg_value(%)  \
0                  product_id    object          0.000           0.0   
1       product_category_name    object          1.851           0.0   
2         product_name_lenght   float64          1.851           0.0   
3  product_description_lenght   float64          1.851           0.0   
4          product_photos_qty   float64          1.851           0.0   
5            product_weight_g   float64          0.006           0.0   
6           product_length_cm   float64          0.006           0.0   
7           product_height_cm   float64          0.006           0.0   
8            product_width_cm   float64          0.006           0.0   

   0_value(%)  duplicate  n_unique  \
0       0.000          0     32951   
1       0.000          0        73   
2       0.000          0        66   
3       0.000          0      2960   
4       0.000          0        19   
5       0.012          0      2204   
6       0.000          0        99   
7       0.000          0       102   
8       0.000          0        95   

                                       sample_unique  
0  [1e9e8ef04dbcff4541ed26657ea517e5, 3aa071139cb...  
1  [perfumaria, artes, esporte_lazer, bebes, util...  
2  [40.0, 44.0, 46.0, 27.0, 37.0, 60.0, 56.0, 57....  
3  [287.0, 276.0, 250.0, 261.0, 402.0, 745.0, 127...  
4  [1.0, 4.0, 2.0, 3.0, 5.0, 9.0, 6.0, nan, 7.0, ...  
5  [225.0, 1000.0, 154.0, 371.0, 625.0, 200.0, 18...  
6  [16.0, 30.0, 18.0, 26.0, 20.0, 38.0, 70.0, 40....  
7  [10.0, 18.0, 9.0, 4.0, 17.0, 5.0, 24.0, 8.0, 1...  
8  [14.0, 20.0, 15.0, 26.0, 13.0, 11.0, 44.0, 40....

In [52]:
df_products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

- Kiểu dữ liệu product_id, product_category_name là object là hợp lý, các cột còn lại là kiểu float.
- có < 2% giá trị null ở các cột trừ cột product_id
- có < 1% giá trị  = 0 tại cột product_weight
- không có giá trị trùng lặp

In [53]:
df_products.dropna(inplace=True)
df_products.drop(df_products[df_products['product_weight_g']==0].index, inplace=True)

In [54]:
df_products['product_category_name'] = df_products['product_category_name'].str.title()
df_products

product_id              product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                         Perfumaria   
1      3aa071139cb16b67ca9e5dea641aaa2f                              Artes   
2      96bd76ec8810374ed1b65e291975717f                      Esporte_Lazer   
3      cef67bcfe19066a932b7673e239eb23d                              Bebes   
4      9dc1a7de274444849c219cff195d0b71              Utilidades_Domesticas   
...                                 ...                                ...   
32946  a0b7d5a992ccda646f2d34e418fff5a0                   Moveis_Decoracao   
32947  bf4538d88321d0fd4412a93c974510e6  Construcao_Ferramentas_Iluminacao   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    Cama_Mesa_Banho   
32949  83808703fc0706a22e264b9d75f04a2e             Informatica_Acessorios   
32950  106392145fca363410d287a815be6de4                    Cama_Mesa_Banho   

       product_name_lenght  product_description_lenght  product_photos_qty  \
0                     40.0                       287.0                 1.0   
1                     44.0                       276.0                 1.0   
2                     46.0                       250.0                 1.0   
3                     27.0                       261.0                 1.0   
4                     37.0                       402.0                 4.0   
...                    ...                         ...                 ...   
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
0                 225.0               16.0               10.0   
1                1000.0               30.0               18.0   
2                 154.0               18.0                9.0   
3                 371.0               26.0                4.0   
4                 625.0               20.0               17.0   
...                 ...                ...                ...   
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm  
0                  14.0  
1                  20.0  
2                  15.0  
3                  26.0  
4                  13.0  
...                 ...  
32946              40.0  
32947              16.0  
32948              27.0  
32949              20.0  
32950               7.0  

[32336 rows x 9 columns]

In [55]:
overview_data(df_products)

feature data_type  null_value(%)  neg_value(%)  \
0                  product_id    object            0.0           0.0   
1       product_category_name    object            0.0           0.0   
2         product_name_lenght   float64            0.0           0.0   
3  product_description_lenght   float64            0.0           0.0   
4          product_photos_qty   float64            0.0           0.0   
5            product_weight_g   float64            0.0           0.0   
6           product_length_cm   float64            0.0           0.0   
7           product_height_cm   float64            0.0           0.0   
8            product_width_cm   float64            0.0           0.0   

   0_value(%)  duplicate  n_unique  \
0         0.0          0     32336   
1         0.0          0        73   
2         0.0          0        66   
3         0.0          0      2960   
4         0.0          0        19   
5         0.0          0      2201   
6         0.0          0        99   
7         0.0          0       102   
8         0.0          0        95   

                                       sample_unique  
0  [1e9e8ef04dbcff4541ed26657ea517e5, 3aa071139cb...  
1  [Perfumaria, Artes, Esporte_Lazer, Bebes, Util...  
2  [40.0, 44.0, 46.0, 27.0, 37.0, 60.0, 56.0, 57....  
3  [287.0, 276.0, 250.0, 261.0, 402.0, 745.0, 127...  
4  [1.0, 4.0, 2.0, 3.0, 5.0, 9.0, 6.0, 7.0, 12.0,...  
5  [225.0, 1000.0, 154.0, 371.0, 625.0, 200.0, 18...  
6  [16.0, 30.0, 18.0, 26.0, 20.0, 38.0, 70.0, 40....  
7  [10.0, 18.0, 9.0, 4.0, 17.0, 5.0, 24.0, 8.0, 1...  
8  [14.0, 20.0, 15.0, 26.0, 13.0, 11.0, 44.0, 40....

In [56]:
overview_data(df_sellers)

feature data_type  null_value(%)  neg_value(%)  0_value(%)  \
0               seller_id    object            0.0           0.0         0.0   
1  seller_zip_code_prefix     int64            0.0           0.0         0.0   
2             seller_city    object            0.0           0.0         0.0   
3            seller_state    object            0.0           0.0         0.0   

   duplicate  n_unique                                      sample_unique  
0          0      3095  [3442f8959a84dea7ee197c632cb2df15, d1b65fc7deb...  
1          0      2246  [13023, 13844, 20031, 4195, 12914, 20920, 5532...  
2          0       611  [campinas, mogi guacu, rio de janeiro, sao pau...  
3          0        23  [SP, RJ, PE, PR, GO, SC, BA, DF, RS, MG, RN, M...

In [57]:
df_sellers.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

- Có 2 kiểu dữ liệu chính là object và int
- không có dup, null, <=0.

In [58]:
filtered_city(df_sellers, 'seller_city')

seller_id  seller_zip_code_prefix  \
237   c3aad7dc65449ae90a5e9c3c6c1e78e0                   15350   
246   71593c7413973a1e160057b80d4958f6                    3407   
360   a3fa18b3f688ec0fca3eb8bfcbd2d5b3                    4557   
476   26b482dccfa29bd2e40703ba45523702                   13450   
551   723a46b89fd5c3ed78ccdf039e33ac63                   93310   
622   7994b065a7ffb14e71c6312cf87b9de2                   29142   
869   cbf09e831b0c11f6f23ffb51004db972                    9726   
874   4aba391bc3b88717ce08eb11e44937b2                   45816   
945   f52c2422904463fdd7741f99045fecb6                    9230   
1004  1cbd32d00d01bb8087a5eb088612fd9c                    3363   
1159  89dda63a3c907c468ec88c310ed91213                    9380   
1337  720e6cf846ea7572cbb66b743fb91e6c                    8717   
1346  cf1313c6e2c01c2f4b014f97db4bcd2b                   22050   
1447  fe9d9cf8631285d5982c6e2cf27fb114                   36200   
1649  d79e8478eed9999493990b44955fb22e                   20081   
1712  6025c79c035c3d772133b8b8238463b2                   83327   
1920  01fd077212124329bac32490e8ef80d9                   14079   
2162  a1bea7061f61f6fdd9a85a6325ba1033                    6311   
2258  4b5f66b7adcf57f1ecc0d3c07dd6b177                   87025   
2340  5dceca129747e92ff8ef7a997dc4f8ca                   13450   
2620  6a3139c7bf09ece22a4713d956acbe5e                   13790   
2988  f9eedec3129e8cc6b6429c42d0808c5b                   22793   
3015  7f5e4d5efad7e44b91115dd1decb65f3                   12306   

                                   seller_city seller_state  
237                               auriflama/sp           SP  
246                      sao paulo / sao paulo           SP  
360                                 são paulo           SP  
476                      santa barbara d´oeste           SP  
551   novo hamburgo, rio grande do sul, brasil           RS  
622                             cariacica / es           ES  
869                                     sbc/sp           SP  
874             arraial d'ajuda (porto seguro)           BA  
945                      santo andre/sao paulo           SP  
1004                                   sp / sp           SP  
1159                            maua/sao paulo           SP  
1337                      mogi das cruzes / sp           SP  
1346            rio de janeiro \rio de janeiro           RJ  
1447                   barbacena/ minas gerais           MG  
1649           rio de janeiro / rio de janeiro           RJ  
1712                                pinhais/pr           PR  
1920                ribeirao preto / sao paulo           SP  
2162                   carapicuiba / sao paulo           SP  
2258                 vendas@creditparts.com.br           PR  
2340                     santa barbara d´oeste           SP  
2620                 sao sebastiao da grama/sp           SP  
2988    rio de janeiro, rio de janeiro, brasil           RJ  
3015                       jacarei / sao paulo           SP

In [59]:
df_sellers['seller_city'] = df_sellers['seller_city'].apply(replace_char)

In [60]:
df_sellers['seller_city'].value_counts()

seller_city
sao paulo              694
curitiba               127
rio de janeiro          96
belo horizonte          68
ribeirao preto          52
                      ... 
taruma                   1
s jose do rio preto      1
domingos martins         1
messias targino          1
leme                     1
Name: count, Length: 611, dtype: int64

In [61]:
filtered_city(df_sellers, 'seller_city')

seller_id  seller_zip_code_prefix  \
237   c3aad7dc65449ae90a5e9c3c6c1e78e0                   15350   
246   71593c7413973a1e160057b80d4958f6                    3407   
360   a3fa18b3f688ec0fca3eb8bfcbd2d5b3                    4557   
476   26b482dccfa29bd2e40703ba45523702                   13450   
551   723a46b89fd5c3ed78ccdf039e33ac63                   93310   
622   7994b065a7ffb14e71c6312cf87b9de2                   29142   
869   cbf09e831b0c11f6f23ffb51004db972                    9726   
874   4aba391bc3b88717ce08eb11e44937b2                   45816   
945   f52c2422904463fdd7741f99045fecb6                    9230   
1004  1cbd32d00d01bb8087a5eb088612fd9c                    3363   
1159  89dda63a3c907c468ec88c310ed91213                    9380   
1337  720e6cf846ea7572cbb66b743fb91e6c                    8717   
1346  cf1313c6e2c01c2f4b014f97db4bcd2b                   22050   
1447  fe9d9cf8631285d5982c6e2cf27fb114                   36200   
1649  d79e8478eed9999493990b44955fb22e                   20081   
1712  6025c79c035c3d772133b8b8238463b2                   83327   
1920  01fd077212124329bac32490e8ef80d9                   14079   
2162  a1bea7061f61f6fdd9a85a6325ba1033                    6311   
2258  4b5f66b7adcf57f1ecc0d3c07dd6b177                   87025   
2340  5dceca129747e92ff8ef7a997dc4f8ca                   13450   
2620  6a3139c7bf09ece22a4713d956acbe5e                   13790   
2988  f9eedec3129e8cc6b6429c42d0808c5b                   22793   
3015  7f5e4d5efad7e44b91115dd1decb65f3                   12306   

                                   seller_city seller_state  
237                               auriflama/sp           SP  
246                      sao paulo / sao paulo           SP  
360                                 são paulo           SP  
476                      santa barbara d´oeste           SP  
551   novo hamburgo, rio grande do sul, brasil           RS  
622                             cariacica / es           ES  
869                                     sbc/sp           SP  
874             arraial d'ajuda (porto seguro)           BA  
945                      santo andre/sao paulo           SP  
1004                                   sp / sp           SP  
1159                            maua/sao paulo           SP  
1337                      mogi das cruzes / sp           SP  
1346            rio de janeiro \rio de janeiro           RJ  
1447                   barbacena/ minas gerais           MG  
1649           rio de janeiro / rio de janeiro           RJ  
1712                                pinhais/pr           PR  
1920                ribeirao preto / sao paulo           SP  
2162                   carapicuiba / sao paulo           SP  
2258                 vendas@creditparts.com.br           PR  
2340                     santa barbara d´oeste           SP  
2620                 sao sebastiao da grama/sp           SP  
2988    rio de janeiro, rio de janeiro, brasil           RJ  
3015                       jacarei / sao paulo           SP

In [62]:
df_sellers = df_sellers.merge(df_geolocation[['geolocation_zip_code_prefix', 'geolocation_city']],
                             left_on= 'seller_zip_code_prefix',
                              right_on= 'geolocation_zip_code_prefix',
                              how='left'
                             ).drop_duplicates()

In [63]:
df_sellers['geolocation_city'] = df_sellers['geolocation_city'].fillna(df_sellers['seller_city'])
df_sellers['geolocation_city'] = df_sellers['geolocation_city'].str.title()
df_sellers['seller_city'] = df_sellers['geolocation_city']
df_sellers.drop(columns=['geolocation_zip_code_prefix', 'geolocation_city'], inplace=True)
df_sellers.head()

seller_id  seller_zip_code_prefix  \
0    3442f8959a84dea7ee197c632cb2df15                   13023   
58   d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
149  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
247  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
307  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

           seller_city seller_state  
0             Campinas           SP  
58          Mogi Guacu           SP  
149     Rio De Janeiro           RJ  
247          Sao Paulo           SP  
307  Braganca Paulista           SP

In [64]:
overview_data(df_cat)

feature data_type  null_value(%)  neg_value(%)  \
0          product_category_name    object            0.0             0   
1  product_category_name_english    object            0.0             0   

   0_value(%)  duplicate  n_unique  \
0           0          0        71   
1           0          0        71   

                                       sample_unique  
0  [beleza_saude, informatica_acessorios, automot...  
1  [health_beauty, computers_accessories, auto, b...

- Kiểu dữ liệu ổn
- không có null, dup, <=0

In [65]:
df_cat.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [66]:
df_cat['product_category_name']=df_cat['product_category_name'].str.title()
df_cat['product_category_name_english']=df_cat['product_category_name_english'].str.title()
df_cat.head()

product_category_name product_category_name_english
0            Beleza_Saude                 Health_Beauty
1  Informatica_Acessorios         Computers_Accessories
2              Automotivo                          Auto
3         Cama_Mesa_Banho                Bed_Bath_Table
4        Moveis_Decoracao               Furniture_Decor

In [67]:
Products = pd.merge(df_products, df_cat, on = 'product_category_name', how = 'left')
Products.drop(columns=['product_category_name'], inplace=True)
Products.rename(columns={'product_category_name_english': 'product_category_name'}, inplace=True)
moving_column = Products.pop('product_category_name')
Products.insert(1,'product_category_name', moving_column)

Products

product_id      product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                  Perfumery   
1      3aa071139cb16b67ca9e5dea641aaa2f                        Art   
2      96bd76ec8810374ed1b65e291975717f             Sports_Leisure   
3      cef67bcfe19066a932b7673e239eb23d                       Baby   
4      9dc1a7de274444849c219cff195d0b71                 Housewares   
...                                 ...                        ...   
32331  a0b7d5a992ccda646f2d34e418fff5a0            Furniture_Decor   
32332  bf4538d88321d0fd4412a93c974510e6  Construction_Tools_Lights   
32333  9a7c6041fa9592d9d9ef6cfe62a71f8c             Bed_Bath_Table   
32334  83808703fc0706a22e264b9d75f04a2e      Computers_Accessories   
32335  106392145fca363410d287a815be6de4             Bed_Bath_Table   

       product_name_lenght  product_description_lenght  product_photos_qty  \
0                     40.0                       287.0                 1.0   
1                     44.0                       276.0                 1.0   
2                     46.0                       250.0                 1.0   
3                     27.0                       261.0                 1.0   
4                     37.0                       402.0                 4.0   
...                    ...                         ...                 ...   
32331                 45.0                        67.0                 2.0   
32332                 41.0                       971.0                 1.0   
32333                 50.0                       799.0                 1.0   
32334                 60.0                       156.0                 2.0   
32335                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
0                 225.0               16.0               10.0   
1                1000.0               30.0               18.0   
2                 154.0               18.0                9.0   
3                 371.0               26.0                4.0   
4                 625.0               20.0               17.0   
...                 ...                ...                ...   
32331           12300.0               40.0               40.0   
32332            1700.0               16.0               19.0   
32333            1400.0               27.0                7.0   
32334             700.0               31.0               13.0   
32335            2083.0               12.0                2.0   

       product_width_cm  
0                  14.0  
1                  20.0  
2                  15.0  
3                  26.0  
4                  13.0  
...                 ...  
32331              40.0  
32332              16.0  
32333              27.0  
32334              20.0  
32335               7.0  

[32336 rows x 9 columns]

In [68]:
df_orders.head(2)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   

  order_status order_purchase_timestamp   order_approved_at  \
0    Delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    Delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13

In [69]:
df_order_items.head(2)

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   

  shipping_limit_date  price  freight_value  
0 2017-09-19 09:45:35   58.9          13.29  
1 2017-05-03 11:05:13  239.9          19.93

In [70]:
Orders = pd.merge(df_orders, df_order_items, on='order_id', how='left')
Orders = Orders.merge(df_order_payments, on='order_id', how='left')
Orders = Orders.merge(df_order_reviews, on='order_id', how='left')
Orders

order_id                       customer_id  \
0       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3       53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4       47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
...                                  ...                               ...   
108429  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
108430  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
108431  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
108432  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
108433  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

       order_status order_purchase_timestamp   order_approved_at  \
0         Delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1         Delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
2         Delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
3         Delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
4         Delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
...             ...                      ...                 ...   
108429    Delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
108430    Delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
108431    Delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
108432    Delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
108433    Delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2017-10-04 19:55:00           2017-10-10 21:25:13   
1               2017-10-04 19:55:00           2017-10-10 21:25:13   
2               2017-10-04 19:55:00           2017-10-10 21:25:13   
3               2018-07-26 14:31:00           2018-08-07 15:27:45   
4               2018-08-08 13:50:00           2018-08-17 18:06:29   
...                             ...                           ...   
108429          2018-02-07 23:22:42           2018-02-28 17:37:56   
108430          2017-08-28 20:52:26           2017-09-21 11:24:17   
108431          2018-01-12 15:35:03           2018-01-25 23:32:54   
108432          2018-01-12 15:35:03           2018-01-25 23:32:54   
108433          2018-03-09 22:11:59           2018-03-16 13:08:30   

       order_estimated_delivery_date  order_item_id  \
0                         2017-10-18            1.0   
1                         2017-10-18            1.0   
2                         2017-10-18            1.0   
3                         2018-08-13            1.0   
4                         2018-09-04            1.0   
...                              ...            ...   
108429                    2018-03-02            1.0   
108430                    2017-09-27            1.0   
108431                    2018-02-15            1.0   
108432                    2018-02-15            2.0   
108433                    2018-04-03            1.0   

                              product_id  ... payment_sequential payment_type  \
0       87285b34884572647811a353c7ac498a  ...                1.0  credit_card   
1       87285b34884572647811a353c7ac498a  ...                3.0      voucher   
2       87285b34884572647811a353c7ac498a  ...                2.0      voucher   
3       595fac2a385ac33a80bd5114aec74eb8  ...                1.0       boleto   
4       aa4383b373c6aca5d8797843e5594415  ...                1.0  credit_card   
...                                  ...  ...                ...          ...   
108429  f1d4ce8c6dd66c47bbaa8c6781c2a923  ...                1.0  credit_card   
108430  b80910977a37536adeddd63663f916ad  ...                1.0  credit_card   
108431  d1c427060a0f73f6b889a5c7c61f2ac4  ...              

In [71]:
overview_data(Orders)

feature       data_type  null_value(%)  \
0                        order_id          object          0.000   
1                     customer_id          object          0.000   
2                    order_status          object          0.000   
3        order_purchase_timestamp  datetime64[ns]          0.000   
4               order_approved_at  datetime64[ns]          0.149   
5    order_delivered_carrier_date  datetime64[ns]          1.924   
6   order_delivered_customer_date  datetime64[ns]          3.146   
7   order_estimated_delivery_date  datetime64[ns]          0.000   
8                   order_item_id         float64          0.768   
9                      product_id          object          0.768   
10                      seller_id          object          0.768   
11            shipping_limit_date  datetime64[ns]          0.768   
12                          price         float64          0.768   
13                  freight_value         float64          0.768   
14             payment_sequential         float64          0.005   
15                   payment_type          object          0.005   
16           payment_installments         float64          0.005   
17                  payment_value         float64          0.005   
18                      review_id          object          0.721   
19                   review_score         float64          0.721   
20           review_comment_title          object         88.526   
21         review_comment_message          object         58.856   
22           review_creation_date  datetime64[ns]          0.721   
23        review_answer_timestamp  datetime64[ns]          0.721   

    neg_value(%)  0_value(%)  duplicate  n_unique  \
0            0.0       0.000          0     90252   
1            0.0       0.000          0     90252   
2            0.0       0.000          0         8   
3            0.0       0.000          0     89789   
4            0.0       0.000          0     83262   
5            0.0       0.000          0     73952   
6            0.0       0.000          0     86599   
7            0.0       0.000          0       458   
8            0.0       0.000          0        20   
9            0.0       0.000          0     31120   
10           0.0       0.000          0      3036   
11           0.0       0.000          0     85296   
12           0.0       0.000          0      5729   
13           0.0       0.336          0      6719   
14           0.0       0.000          0        29   
15           0.0       0.000          0         4   
16           0.0       0.000          0        22   
17           0.0       0.006          0     27771   
18           0.0       0.000          0     89457   
19           0.0       0.000          0         5   
20           0.0       0.000          0      4026   
21           0.0       0.000          0     31857   
22           0.0       0.000          0       635   
23           0.0       0.000          0     89325   

                                        sample_unique  
0   [e481f51cbdc54678b7cc49136f2d6af7, 53cdb2fc8bc...  
1   [9ef432eb6251297304e76186b10a928d, b0830fb4747...  
2   [Delivered, Invoiced, Shipped, Processing, Una...  
3   [2017-10-02 10:56:33, 2018-07-24 20:41:37, 201...  
4   [2017-10-02 11:07:15, 2018-07-26 03:24:27, 201...  
5   [2017-10-04 19:55:00, 2018-07-26 14:31:00, 201...  
6   [2017-10-10 21:25:13, 2018-08-07 15:27:45, 201...  
7   [2017-10-18 00:00:00, 2018-08-13 00:00:00, 201...  
8   [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, nan, 7.0, 8.0, ...  
9   [87285b34884572647811a353c7ac498a, 595fac2a385...  
10  [3504c0cb71d7fa48d967e0e4c94d59d9, 289cdb325fb...  
11  [2017-10-06 11:07:15, 2018-07-30 03:24:27, 201...  
12  [29.99, 118.7, 159.9, 45.0, 19.9, 147.9, 49.9,...  
13  [8.72, 22.76, 19.22, 27.2, 27.36, 16.05, 15.17...  
14  [1.0, 3.0, 2.0, 4.0, nan, 8.0, 5.0, 7.0, 9.0, ...  
15    [credit_card, voucher, boleto, debit_card, nan]  
16  [1.0, 3.0, 6.0, 10.0, 4.0, 2.0, 8.0, 9.0, 7

In [72]:
overview_data(Orders)

feature       data_type  null_value(%)  \
0                        order_id          object          0.000   
1                     customer_id          object          0.000   
2                    order_status          object          0.000   
3        order_purchase_timestamp  datetime64[ns]          0.000   
4               order_approved_at  datetime64[ns]          0.149   
5    order_delivered_carrier_date  datetime64[ns]          1.924   
6   order_delivered_customer_date  datetime64[ns]          3.146   
7   order_estimated_delivery_date  datetime64[ns]          0.000   
8                   order_item_id         float64          0.768   
9                      product_id          object          0.768   
10                      seller_id          object          0.768   
11            shipping_limit_date  datetime64[ns]          0.768   
12                          price         float64          0.768   
13                  freight_value         float64          0.768   
14             payment_sequential         float64          0.005   
15                   payment_type          object          0.005   
16           payment_installments         float64          0.005   
17                  payment_value         float64          0.005   
18                      review_id          object          0.721   
19                   review_score         float64          0.721   
20           review_comment_title          object         88.526   
21         review_comment_message          object         58.856   
22           review_creation_date  datetime64[ns]          0.721   
23        review_answer_timestamp  datetime64[ns]          0.721   

    neg_value(%)  0_value(%)  duplicate  n_unique  \
0            0.0       0.000          0     90252   
1            0.0       0.000          0     90252   
2            0.0       0.000          0         8   
3            0.0       0.000          0     89789   
4            0.0       0.000          0     83262   
5            0.0       0.000          0     73952   
6            0.0       0.000          0     86599   
7            0.0       0.000          0       458   
8            0.0       0.000          0        20   
9            0.0       0.000          0     31120   
10           0.0       0.000          0      3036   
11           0.0       0.000          0     85296   
12           0.0       0.000          0      5729   
13           0.0       0.336          0      6719   
14           0.0       0.000          0        29   
15           0.0       0.000          0         4   
16           0.0       0.000          0        22   
17           0.0       0.006          0     27771   
18           0.0       0.000          0     89457   
19           0.0       0.000          0         5   
20           0.0       0.000          0      4026   
21           0.0       0.000          0     31857   
22           0.0       0.000          0       635   
23           0.0       0.000          0     89325   

                                        sample_unique  
0   [e481f51cbdc54678b7cc49136f2d6af7, 53cdb2fc8bc...  
1   [9ef432eb6251297304e76186b10a928d, b0830fb4747...  
2   [Delivered, Invoiced, Shipped, Processing, Una...  
3   [2017-10-02 10:56:33, 2018-07-24 20:41:37, 201...  
4   [2017-10-02 11:07:15, 2018-07-26 03:24:27, 201...  
5   [2017-10-04 19:55:00, 2018-07-26 14:31:00, 201...  
6   [2017-10-10 21:25:13, 2018-08-07 15:27:45, 201...  
7   [2017-10-18 00:00:00, 2018-08-13 00:00:00, 201...  
8   [1.0, 2.0, 3.0, 4.0, 5.0, 6.0, nan, 7.0, 8.0, ...  
9   [87285b34884572647811a353c7ac498a, 595fac2a385...  
10  [3504c0cb71d7fa48d967e0e4c94d59d9, 289cdb325fb...  
11  [2017-10-06 11:07:15, 2018-07-30 03:24:27, 201...  
12  [29.99, 118.7, 159.9, 45.0, 19.9, 147.9, 49.9,...  
13  [8.72, 22.76, 19.22, 27.2, 27.36, 16.05, 15.17...  
14  [1.0, 3.0, 2.0, 4.0, nan, 8.0, 5.0, 7.0, 9.0, ...  
15    [credit_card, voucher, boleto, debit_card, nan]  
16  [1.0, 3.0, 6.0, 10.0, 4.0, 2.0, 8.0, 9.0, 7

# Write File

In [73]:
path_des = r"C:\Users\ADMIN\Downloads\Project\Brazillian E_Commerce\Data Cleaned\\"
df_customer.to_csv(path_des + "Customer.csv", index=False)
df_geolocation.to_csv(path_des + "Geolocation.csv", index=False)
df_orders.to_csv(path_des + "Orders.csv", index=False)
df_order_items.to_csv(path_des + "Order_items.csv", index=False)
df_order_payments.to_csv(path_des + "Order_payments.csv", index=False)
df_order_reviews.to_csv(path_des + "Order_reviews.csv", index=False)
df_products.to_csv(path_des + "Product.csv", index=False)
df_sellers.to_csv(path_des + "Sellers.csv", index=False)
df_cat.to_csv(path_des + "Cat.csv", index=False)
Orders.to_csv(path_des + "Orders.csv", index=False)
Products.to_csv(path_des + "Products.csv", index=False)

In [74]:
df_customer

key_0                       customer_id  \
0         14409  06b8999e2fba1a1fbc88172c00ba8bc7   
126        9790  18955e83d337fd6b2def6b18a428ac77   
251        1151  4e7b3e00288586ebd08712fdd0374a03   
285        8775  b2b6027bc5c5109e529d4dc6358b12c3   
368       13056  4f2d8ab171c80ec8364f7c12e35b23ad   
...         ...                               ...   
10327568   3937  17ddf5dd5d51696bb3d7c6291687be6f   
10327605   6764  e7b71a9017aa05c9a7fd292d714858e8   
10327685  60115  5e28dfe12db7fb50a4b2f691faecea5e   
10327784  92120  56b18e2166679b8a959d72dd06da27f9   
10327898   6703  274fa6071e5e17fe303b9748641082c8   

                        customer_unique_id  customer_zip_code_prefix  \
0         861eff4711a542e4b93843c6dd7febb0                     14409   
126       290c77bc529b7ac935b93aa66c333dc3                      9790   
251       060e732b5b29e8181a18229c7b0b2b5e                      1151   
285       259dac757896d24d7702b9acbbff3f3c                      8775   
368       345ecd01c38d18a9036ed96c73b8d066                     13056   
...                                    ...                       ...   
10327568  1a29b476fee25c95fbafc67c5ac95cf8                      3937   
10327605  d52a67c98be1cf6a5c84435bd38d095d                      6764   
10327685  e9f50caf99f032f0bf3c55141f019d99                     60115   
10327784  73c2643a0a458b49f58cea58833b192e                     92120   
10327898  84732c5050c01db9b23e19ba39899398                      6703   

                  customer_city customer_state  
0                        Franca             SP  
126       Sao Bernardo Do Campo             SP  
251                   Sao Paulo             SP  
285             Mogi Das Cruzes             SP  
368                    Campinas             SP  
...                         ...            ...  
10327568              Sao Paulo             SP  
10327605        Taboao Da Serra             SP  
10327685              Fortaleza             CE  
10327784                 Canoas             RS  
10327898                  Cotia             SP  

[99441 rows x 6 columns]

In [75]:
Products.dtypes

product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object